rust/cubestore/src/queryplanner/udfs.rs

```diff
@@ -180,6 +189,108 @@ impl CubeScalarUDF for UnixTimestamp {
     }
 }
 
+struct DateAdd {}
+impl DateAdd {
+    fn signature() -> Signature {
+        Signature::Any(2)
+    }
+}
+impl CubeScalarUDF for DateAdd {
+    fn kind(&self) -> CubeScalarUDFKind {
+        CubeScalarUDFKind::DateAdd
+    }
+
+    fn name(&self) -> &str {
+        "DATE_ADD"
+    }
+
+    fn descriptor(&self) -> ScalarUDF {
+        return ScalarUDF {
+            name: self.name().to_string(),
+            signature: Self::signature(),
+            return_type: Arc::new(|inputs| {
+                assert!(inputs.len() == 2);
+
+                // Right now, We support only TimeUnit::Nanosecond without TZ
+                Ok(Arc::new(DataType::Timestamp(TimeUnit::Nanosecond, None)))
+            }),
+            fun: Arc::new(|inputs| {
+                if inputs.len() != 2 {
+                    return Err(DataFusionError::Plan(
+                        "Expected two arguments in DATE_ADD".to_string(),
+                    ));
+                }
+
+                let mut result_date = match &inputs[0] {
+                    ColumnarValue::Scalar(scalar) => match (scalar.get_datatype(), scalar) {
+                        // Right now, We support only TimeUnit::Nanosecond without TZ
+                        (
+                            DataType::Timestamp(TimeUnit::Nanosecond, None),
+                            ScalarValue::TimestampNanosecond(Some(v)),
+                        ) => Utc.timestamp_nanos(*v),
+                        _ => {
+                            return Err(DataFusionError::Plan(
+                                "First argument of `DATE_PART` must be non-null scalar TimestampNanosecond without timezone"
+                                    .to_string(),
+                            ));
+                        }
+                    },
+                    _ => {
+                        return Err(DataFusionError::Plan(
+                            "First argument of `DATE_PART` must be non-null scalar TimestampNanosecond without timezone"
+                                .to_string(),
+                        ));
+                    }
+                };
+
+                match &inputs[1] {
+                    ColumnarValue::Scalar(scalar) => match scalar {
+                        ScalarValue::IntervalYearMonth(Some(v)) => {
+                            let years_to_add = (*v as f64 / 12_f64).floor() as i32;
+                            let months_to_add = (*v - (years_to_add * 12)) as u32;
```

Comment:
nit: replace with `*v % 12`. It's shorter.

Reply:
✅